src/ocf_seq_cutoff.c

```diff
@@ -146,55 +205,130 @@ bool ocf_core_seq_cutoff_check(ocf_core_t core, struct ocf_request *req)
+		struct ocf_seq_cutoff *seq_cutoff,
+		uint64_t addr, uint32_t len, int rw, bool insert)
 {
-	ocf_seq_cutoff_policy policy = ocf_core_get_seq_cutoff_policy(core);
 	struct ocf_seq_cutoff_stream item = {
-		.last = req->byte_position, .rw = req->rw
+		.last = addr, .rw = rw
 	};
 	struct ocf_seq_cutoff_stream *stream;
 	struct ocf_rb_node *node;
 	bool can_update;
 
-	if (policy == ocf_seq_cutoff_policy_never)
-		return;
-
-	/* Update last accessed position and bytes counter */
-	env_rwlock_write_lock(&core->seq_cutoff->lock);
-	node = ocf_rb_tree_find(&core->seq_cutoff->tree, &item.node);
+	node = ocf_rb_tree_find(&seq_cutoff->tree, &item.node);
 	if (node) {
 		stream = container_of(node, struct ocf_seq_cutoff_stream, node);
-		item.last = req->byte_position + req->byte_length;
-		can_update = ocf_rb_tree_can_update(&core->seq_cutoff->tree,
+		item.last = addr + len;
+		can_update = ocf_rb_tree_can_update(&seq_cutoff->tree,
 				node, &item.node);
-		stream->last = req->byte_position + req->byte_length;
-		stream->bytes += req->byte_length;
+		stream->last = addr + len;
+		stream->bytes += len;
+		stream->req_count++;
 		if (!can_update) {
-			ocf_rb_tree_remove(&core->seq_cutoff->tree, node);
-			ocf_rb_tree_insert(&core->seq_cutoff->tree, node);
+			ocf_rb_tree_remove(&seq_cutoff->tree, node);
+			ocf_rb_tree_insert(&seq_cutoff->tree, node);
 		}
-		list_move_tail(&stream->list, &core->seq_cutoff->lru);
-	} else {
-		stream = list_first_entry(&core->seq_cutoff->lru,
+		list_move_tail(&stream->list, &seq_cutoff->lru);
+
+		return stream;
+	}
+
+	if (insert) {
+		stream = list_first_entry(&seq_cutoff->lru,
 				struct ocf_seq_cutoff_stream, list);
-		ocf_rb_tree_remove(&core->seq_cutoff->tree, &stream->node);
-		stream->rw = req->rw;
-		stream->last = req->byte_position + req->byte_length;
-		stream->bytes = req->byte_length;
-		ocf_rb_tree_insert(&core->seq_cutoff->tree, &stream->node);
-		list_move_tail(&stream->list, &core->seq_cutoff->lru);
+		ocf_rb_tree_remove(&seq_cutoff->tree, &stream->node);
+		stream->rw = rw;
+		stream->last = addr + len;
+		stream->bytes = len;
+		stream->req_count = 0;
+		stream->valid = true;
+		ocf_rb_tree_insert(&seq_cutoff->tree, &stream->node);
+		list_move_tail(&stream->list, &seq_cutoff->lru);
+
+		return stream;
+	}
+
+	return NULL;
+}
+
+static void ocf_core_seq_cutoff_base_promote(struct ocf_seq_cutoff *seq_cutoff,
+		struct ocf_seq_cutoff_stream *src_stream)
+{
+	struct ocf_seq_cutoff_stream *dst_stream;
+
+	dst_stream = list_first_entry(&seq_cutoff->lru,
+			struct ocf_seq_cutoff_stream, list);
+	ocf_rb_tree_remove(&seq_cutoff->tree, &dst_stream->node);
+	dst_stream->rw = src_stream->rw;
+	dst_stream->last = src_stream->last;
+	dst_stream->bytes = src_stream->bytes;
+	dst_stream->req_count = src_stream->req_count;
+	dst_stream->valid = true;
+	ocf_rb_tree_insert(&seq_cutoff->tree, &dst_stream->node);
+	list_move_tail(&dst_stream->list, &seq_cutoff->lru);
+	src_stream->valid = false;
```

Comment:
how about
list_move_head(&src_stream->list, &<per_queue_cutoff>->lru);
to make sure that the slot after promoted stream in per-queue cutoff is reused asap for a new stream?

Reply:
Fixed.